gatsby-node.js

```diff
@@ -9,45 +9,126 @@ const report = require('gatsby-cli/lib/reporter');
 
+  const indexState = {}
+
   const jobs = queries.map(async function doQuery(
-    { indexName = mainIndexName, query, transformer = identity, settings },
+    { indexName = mainIndexName, query, transformer = identity, settings, matchFields = mainMatchFields },
     i
   ) {
     if (!query) {
       report.panic(
         `failed to index to Algolia. You did not give "query" to this query`
       );
     }
-    const index = client.initIndex(indexName);
-    const mainIndexExists = await indexExists(index);
-    const tmpIndex = client.initIndex(`${indexName}_tmp`);
-    const indexToUse = mainIndexExists ? tmpIndex : index;
+    if (!Array.isArray(matchFields) || !matchFields.length) {
+      return report.panic(
+        `failed to index to Algolia. Argument matchFields has to be an array of strings`
+      );
+    }
 
-    if (mainIndexExists) {
-      setStatus(activity, `query ${i}: copying existing index`);
-      await scopedCopyIndex(client, index, tmpIndex);
+    const index = client.initIndex(indexName);
+    /* Use temp index if main index already exists */
+    let useTempIndex = false
+    const indexToUse = await (async function(_index) {
+      if (!enablePartialUpdates) {
+        if (useTempIndex = await indexExists(_index)) {
+          return client.initIndex(`${indexName}_tmp`);
+        }
+      }
+      return _index
+    })(index)
+
+    /* Use to keep track of what to remove afterwards */
+    if (!indexState[indexName]) indexState[indexName] = {
+      index,
+      toRemove: {}
     }
+    const currentIndexState = indexState[indexName]
 
     setStatus(activity, `query ${i}: executing query`);
     const result = await graphql(query);
     if (result.errors) {
       report.panic(`failed to index to Algolia`, result.errors);
     }
     const objects = transformer(result);
-    const chunks = chunk(objects, chunkSize);
+
+    if (objects.length > 0 && !objects[0].objectID) {
+      report.panic(
+        `failed to index to Algolia. Query results do not have 'objectID' key`
+      );
+    }
+
+    setStatus(activity, `query ${i}: graphql resulted in ${Object.keys(objects).length} records`);
+
+    let hasChanged = objects;
+    let algoliaObjects = {}
+    if (enablePartialUpdates) {
+      setStatus(activity, `query ${i}: starting Partial updates`);
+
+      algoliaObjects = await fetchAlgoliaObjects(indexToUse, matchFields);
+
+      const results = Object.keys(algoliaObjects).length
+      setStatus(activity, `query ${i}: found ${results} existing records`);
+
+      if (results) {
+        hasChanged = objects.filter(curObj => {
+          const ID = curObj.objectID
+          let extObj = algoliaObjects[ID]
+
+          /* The object exists so we don't need to remove it from Algolia */
+          delete(algoliaObjects[ID]);
+          delete(currentIndexState.toRemove[ID])
+
+          if (!extObj) return true;
+
+          return !!matchFields.find(field => extObj[field] !== curObj[field]);
```

Comment:
does this do shallow comparison only? IE what happens for nested fields?

Reply:
Yes shallow comparison only.

Reply:
any potential ideas for nested comparisons? worth mentioning in readme?